hokusai/lib/global_config.py

```diff
@@ -1,54 +1,69 @@
-from packaging.version import Version, InvalidVersion
-
-from hokusai import CWD
+from hokusai.lib.common import print_red
+from hokusai.lib.config_loader import ConfigLoader
 from hokusai.lib.constants import YAML_HEADER
 from hokusai.lib.exceptions import HokusaiError
-from hokusai.version import VERSION
 
-HOKUSAI_GLOBAL_CONFIG_FILE = os.path.join(os.environ.get('HOME', '/'), '.hokusai', 'config.yml')
-
-class HokusaiGlobalConfig:
-  def is_present(self):
-    return os.path.isfile(HOKUSAI_GLOBAL_CONFIG_FILE)
 
-  def get(self, key, default=None, use_env=False, _type=str):
-    value = self._config_value_for(key, _type)
-    if value is not None:
-      return value
+HOKUSAI_GLOBAL_CONFIG_FILE = os.path.join(os.environ.get('HOME', '/'), '.hokusai.yml')
 
-    return default
+class HokusaiGlobalConfig:
+  ''' manage Hokusai global config '''
+  def __init__(self, uri=HOKUSAI_GLOBAL_CONFIG_FILE):
+    self._config = ConfigLoader(uri).load()
+    self.validate_config()
 
+  def merge(self, **kwargs):
+    ''' merge key/value pairs into config, skipping None '''
+    for k,v in kwargs.items():
+      if v is not None:
+        # use dash in _config keys
+        self._config[k.replace('_', '-')] = v
 
-  def _config_value_for(self, key, _type):
+  def save(self):
+    ''' save config to local config file '''
     try:
-      with open(HOKUSAI_GLOBAL_CONFIG_FILE, 'r') as config_file:
-        config_struct = yaml.safe_load(config_file.read())
-        try:
-          val = config_struct[key]
-        except KeyError:
-          return None
-        if not isinstance(val, _type):
-          raise HokusaiError("Config key %s is not of %s" % (key, _type))
-        return val
-    except IOError:
-      return None
+      with open(HOKUSAI_GLOBAL_CONFIG_FILE, 'w') as output:
+        output.write(YAML_HEADER)
+        yaml.safe_dump(
+          self._config,
+          output,
+          default_flow_style=False
+        )
+    except:
+      print_red(
+        f'Error: Not able to write Hokusai config to {file_path}'
+      )
+      raise
 
+  def validate_config(self):
+    ''' sanity check config '''
+    required_vars = [
+      'kubectl-version',
+      'kubeconfig-dir',
+      'kubeconfig-source-uri',
+      'kubectl-dir'
+    ]
+    for var in required_vars:
+      if not var in self._config:
+        raise HokusaiError(
+          f'{var} is missing in Hokusai global config'
+        )
 
   @property
-  def kubectl_version(self):
-    return self.get('kubectl-version')
+  def kubeconfig_dir(self):
+    return self._config['kubeconfig-dir']
 
   @property
-  def kubectl_config_file(self):
-    return self.get('kubectl-config-file')
+  def kubeconfig_source_uri(self):
+    return self._config['kubeconfig-source-uri']
 
+  @property
+  def kubectl_dir(self):
+    return self._config['kubectl-dir']
 
-global_config = HokusaiGlobalConfig()
+  @property
+  def kubectl_version(self):
+    return self._config['kubectl-version']
```

Comment:
wonder if this is better called 'k8s-version', looking at it from cluster's perspective, and kubectl should match cluster version anyway.